schema/schema.go

```diff
@@ -0,0 +1,215 @@
+/*
+   Copyright © 2022 The CDI Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package schema
+
+import (
+	"embed"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+	"strings"
+
+	"sigs.k8s.io/yaml"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+	schema "github.com/xeipuuv/gojsonschema"
+)
+
+const (
+	// BuiltinSchemaName references the builtin schema for Load()/Set().
+	BuiltinSchemaName = "builtin"
+	// DefaultSchemaName is the builtin schema.
+	DefaultSchemaName = BuiltinSchemaName
+	// NoneSchemaName references a disabled/NOP schema for Load()/Set().
+	NoneSchemaName = "none"
+
+	// builtinSchemaFile is the builtin schema URI in our embedded FS.
+	builtinSchemaFile = "file:///schema.json"
+)
+
+// Schema is a JSON validation schema.
+type Schema struct {
+	schema *schema.Schema
+}
+
+// Error wraps a JSON validation result.
+type Error struct {
+	Result *schema.Result
+}
+
+// Set sets the default validating JSON schema.
+func Set(s *Schema) error {
+	current = s
+	return nil
+}
+
+// Get returns the active validating JSON schema.
+func Get() *Schema {
+	return current
+}
+
+// BuiltinSchema returns the builtin validating JSON Schema.
+func BuiltinSchema() *Schema {
+	return builtin
+}
+
+// NopSchema returns an validating JSON Schema that does no real validation.
+func NopSchema() *Schema {
+	return &Schema{}
+}
+
+// ValidateData validates the given JSON document against the active schema.
+func ValidateData(data []byte) (error, error) {
+	return current.ValidateData(data)
+}
+
+// ValidateFile validates the given JSON file against the active schema.
+func ValidateFile(path string) (error, error) {
+	return current.ValidateFile(path)
+}
+
+// Load the given JSON Schema.
+func Load(source string) (*Schema, error) {
+	var (
+		loader schema.JSONLoader
+		err    error
+		s      *schema.Schema
+	)
+
+	source = strings.TrimSpace(source)
+
+	switch {
+	case source == BuiltinSchemaName:
+		return BuiltinSchema(), nil
+	case source == NoneSchemaName, source == "":
+		return NopSchema(), nil
+	case strings.HasPrefix(source, "file://"):
+	case strings.HasPrefix(source, "http://"):
+	case strings.HasPrefix(source, "https://"):
+	default:
+		if strings.Index(source, "://") < 0 {
+			source, err = filepath.Abs(source)
+			if err != nil {
+				return nil, err
```

Comment:
Do we need more context in this `err` being returned?

Reply:
Yes, I think we do. Thanks for spotting. Updated with a better error message.